asmcomp/cmmgen.ml

```diff
@@ -1364,9 +1364,17 @@ let rec transl = function
         Cop(Capply(typ_addr, dbg), [get_field clos 0; transl arg; clos]))
   | Ugeneric_apply(clos, args, dbg) ->
       let arity = List.length args in
-      let cargs = Cconst_symbol(apply_function arity) ::
-        List.map transl (args @ [clos]) in
-      Cop(Capply(typ_addr, dbg), cargs)
+      bind "fun" (transl clos) (fun transl_clos ->
+        let transl_args = List.map transl args in
+        let transl_all_args = transl_args @ [transl_clos] in
+        (* Generate a direct call if this is the first and only application
+           of the closure. Otherwise, emit a call to caml_apply. *)
```

Comment:
As a non-expert "the first and only application of the closure" conveys the wrong intuition to me: I think of a function that is only called once in the program, while if I understand correctly you are saying that all parameters are passed at the same time.